src/core/common/settings.cpp

```diff
@@ -119,11 +119,10 @@ otError Settings::SaveOperationalDataset(bool aIsActive, const MeshCoP::Dataset
 
 otError Settings::ReadOperationalDataset(bool aIsActive, MeshCoP::Dataset &aDataset) const
 {
-    otError  error = OT_ERROR_NONE;
-    uint16_t length;
+    otError  error  = OT_ERROR_NONE;
+    uint16_t length = MeshCoP::Dataset::kMaxSize;
 
-    SuccessOrExit(error = Read(aIsActive ? kKeyActiveDataset : kKeyPendingDataset, aDataset.GetBytes(),
-                               MeshCoP::Dataset::kMaxSize, length));
+    SuccessOrExit(error = Read(aIsActive ? kKeyActiveDataset : kKeyPendingDataset, aDataset.GetBytes(), length));
```

Comment:
Just to be safe here, should we do a `VerifyOrExit(length < MeshCoP::Dataset::kMaxSize);`? (or maybe `assert()` instead)?
The `length` upon return from `Read()` would be number of bytes available for the given key (not the number of bytes actually read and copied into the buffer).